chain/chain/src/store.rs

```diff
@@ -234,6 +314,73 @@ impl ChainStore {
             }
         }
     }
+
+    pub fn check_blocks_on_same_chain(
+        &mut self,
+        cur_header: &BlockHeader,
+        base_block_hash: &CryptoHash,
+        max_difference_in_height: u64,
+    ) -> bool {
+        // first step: update cache head
+        if self.header_history.is_empty() {
+            self.header_history.push_back(cur_header.clone());
+        }
+        let mut prev_block_hash = cur_header.inner.prev_hash;
+
+        if self.header_history.contains(&cur_header.hash) {
+            self.header_history.update(&cur_header.hash, vec![]);
+        } else {
+            let mut header_list = vec![cur_header.clone()];
+            let mut find_ancestor = false;
+            while !self.header_history.is_empty() {
+                let prev_block_header = if let Ok(header) = self.get_block_header(&prev_block_hash)
+                {
+                    header.clone()
+                } else {
+                    return false;
+                };
+                if self.header_history.contains(&prev_block_header.hash) {
+                    self.header_history.update(&prev_block_header.hash, header_list.clone());
+                    find_ancestor = true;
+                    break;
+                }
+                self.header_history.pop_front();
+                prev_block_hash = prev_block_header.inner.prev_hash;
+                header_list.push(prev_block_header);
+                if header_list.len() >= max_difference_in_height as usize {
+                    self.header_history = HeaderList::from_headers(header_list.clone());
+                    break;
+                }
+            }
+            if !find_ancestor {
+                self.header_history = HeaderList::from_headers(header_list);
+            }
+        }
+
+        // second step: check if `base_block_hash` exists
+        if self.header_history.contains(base_block_hash) {
+            return true;
+        }
+        assert!(max_difference_in_height >= self.header_history.len() as u64);
```

Comment:
As discussed offline, the misplaced `pop_front` above can result in this assert triggering.
Move the assert above the conditional return above, and make sure there are tests that trigger it, before fixing the `pop_front`.